### DISCUSSION TITLE
Directus, virtual machine, auto logout

### DISCUSSION THREAD

**Discussion by korneliusz-w** [directus/directus#6255] [Archive]
Hi 

I have a problem.
I have fresh Directus 9 runing on virtual machine Ubuntu 18. After log in and changed few tabs application logs me out. After second log in I can not change any tab. Always log out. 

Everything works good on local machine WIndows with normal npx directus start.

Has anyone encountered this error?

**Comment by korneliusz-w** [comment]
Does nodeJS version matter? On windows I have 14.16.0 (works), on VM Ubuntu 14.4.0 (not working), on local Ubuntu 15.9.0 (not working).

**Reply by rijkvanzanten** [reply]
Hard to say! Could it be that whatever webserver you use as reverse proxy on the ubuntu side strips out headers / cookies?

**Reply by korneliusz-w** [reply]
On Ubuntu I dont have any nginx, only fresh directus.

**Reply by hugopvl** [reply]
The virtual machine does not go into sleep mode? or stop some services ?

**Reply by korneliusz-w** [reply]
No.
I found that in Ubuntu there are no cookies in request.
I try to set nginx proxy, still not working.

```
server {
    listen 80 default_server;;
    listen [::]:80 default_server;

    server_name localhost;

    location / {
        proxy_pass http://localhost:8055;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection 'upgrade';
        proxy_set_header Host $host;
   }
}
```

**Reply by rijkvanzanten** [reply]
> I found that in Ubuntu there are no cookies in request.

That sounds like a problem 🤔 Without the cookie, the app won't be able to use the /refresh endpoint and thus log you out as soon as the access token expires (15m by default)

**Comment by philleepflorence** [comment]
Curious to see what the case may be...

- Is `Ubuntu 14.4.0 ` on your local device accessed via the VM IP Address on host browser or VM browser?
    -  `http://10.0.2.1:8055` vs `http://localhost:8055`.
    - In the D9 `.env` do you use `localhost`, `127.0.0.1`, or `0.0.0.0`?
- Are any other Headers showing up in Browser Request and Node console?
- Is the date on the VM always accurate on restart.

<details>
<summary>Sample Request Headers and Server Tips...</summary>
<div>
  
---
### Sample XHR Request

```console
Accept: application/json, text/plain, */*
Pragma: no-cache
Cookie: directus_refresh_token=Ty7qx[...]
Authorization: Bearer eyJhb[...]
Referer: http://localhost:8055/admin/collections/
Cache-Control: no-cache
Host: localhost:8055
Accept-Language: en-us
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko)
Accept-Encoding: gzip, deflate
Connection: keep-alive
```
---
### Remote VM Server Configuration

**NGINX Config**
```shell

server {
	listen 80;
	listen [::]:80;
	
	server_name directus.example.com;
	
	location / {
		proxy_set_header X-Real-IP $remote_addr;
		proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
		proxy_set_header Host $http_host;
		proxy_set_header X-NginX-Proxy true;
		proxy_pass http://127.0.0.1:8055;
		proxy_redirect off;
		proxy_http_version 1.1;
		proxy_set_header Upgrade $http_upgrade;
		proxy_set_header Connection "upgrade";
	}
}
```
---
**Notes**
- If running NGINX on a virtual machine in VirtualBox
    - In nginx.conf, turn off `sendfile off;`.
- Explicitly set `underscores_in_headers on;` if NGINX drops HTTP headers with underscores.
---
</div>
</details>

**Reply by korneliusz-w** [reply]
I resolved.
Directus should be set on https. If not than the POST login request can't set refresh token cookie.
